klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,194 @@
+import os
+import logging
+import threading
+import sys
+
+import dulwich
+
+from klaus.repo import FancyRepo
+from klaus import Klaus, make_app
+
+try:
+    from watchdog.observers import Observer
+    from watchdog.events import FileSystemEventHandler
+except ImportError:
+    sys.stderr.write("To automatically detect and reload repositories as they "
+                     "change, watchdog must be installed.\n")
+    sys.stderr.write("See https://pypi.python.org/pypi/watchdog/\n")
+    sys.exit(1)
+
+logger = logging.getLogger(__name__)
+
+
+def is_subdirectory(haystack, needle):
+    """
+    checks whether `needle` exists as a subdirectory of `haystack`
+    """
+    haystack = os.path.realpath(haystack)
+    needle = os.path.realpath(needle)
+    if not os.path.isdir(needle):
+        return False
+    return needle.startswith(haystack)
+
+
+class KlausFSEventHandler(FileSystemEventHandler):
+    """
+    Basic file-system event handler for a klaus app.
+    When directories are created, moved or deleted, we will be notified here.
+    We try and update the klaus wsgi app with modifications to git repos as
+    they happen.
+    """
+    def __init__(self, klausapp, *args, **kwargs):
+        self._klausapp = klausapp
+        return super(KlausFSEventHandler, self).__init__(*args, **kwargs)
+
+    def on_modified(self, event):
+        if not event.is_directory:
+            return
+        self._klausapp.remove_repo(event.src_path)
+        self._klausapp.add_repo(event.src_path)
+
+    def on_created(self, event):
+        # We're only interested in directories; A git repo is always a directory
+        if not event.is_directory:
+            return
+        logger.debug("Received directory creation event for '%s'", event.src_path)
+        self._klausapp.add_repo(event.src_path)
+
+    def on_deleted(self, event):
+        if not event.is_directory:
+            return
+        logger.debug("Received directory deletion event for '%s'", event.src_path)
+        self._klausapp.remove_repo(event.src_path)
+
+    def on_moved(self, event):
+        """
+        Moved is not the same as deleted, unless the moved dir is no longer a
+        subset of any of the watched trees. In that case we need to consider it
+        a deletion, and remove it from the list of watched repos; otherwise we
+        need to update its path. This is a potentially expensive operation,
+        but shouldn't affect performance in the general case, as moving a
+        repository is a rare event.
+        """
+        logger.debug(
+            "Received repository move event:'%s'->'%s'",
+            event.src_path,
+            event.dest_path
+        )
```

Comment:
Try to keep the visual space occupied by logging lines as small as possible. Condense this to a single line. It's fine if it's 100 chars wide or so.